### PR TITLE
fix(parser): report syntax errors for missing constructor implementations

### PR DESCRIPTION
**Pull request by camc314**

### CODE REVIEW

**Comment by camc314** [comment]
* **#8081** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8081?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8081?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=8081).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: crates/oxc_semantic/src/builder.rs [hunk 1]
```diff
@@ -306,6 +306,14 @@ impl<'a> SemanticBuilder<'a> {
         self.errors.borrow_mut().push(error);
     }
 
+    pub(crate) fn in_declare_scope(&self) -> bool {
+        self.source_type.is_typescript_definition()
+            || self
```
Comment (Boshen) [inline crates/oxc_semantic/src/builder.rs:311]: This is expensive, try use the ambient flag - move the check logic to the parser if semantic don't have this context.

Comment (camc314) [reply crates/oxc_semantic/src/builder.rs:311]: i don't think we've enough info in the parser, but i can try?

what do you think about introducing a self.flags for storing this?
that shouldn't incur the same perf cost

**Comment by Boshen** [comment]
### Merge activity

* **Dec 23, 8:35 PM EST**: [Graphite](https://app.graphite.dev/github/pr/oxc-project/oxc/8081) disabled "merge when ready" on this PR due to: a merge conflict with the target branch; resolve the conflict and try again..

**Comment by camc314** [comment]
will rebase these in the morning

**Pull request merged**